sgx/common/src/ec_key_cert_marshal.c

```diff
@@ -0,0 +1,112 @@
+/**
+ * @file ec_key_cert_marshal.c
+ *
+ * @brief Provides implementation of functionality used to marshal (serialize)
+ *        elliptic curve key and cert structs. This supports passing these
+ *        variables across the enclave boundary, as a binary data format is
+ *        required.
+ */
+
+#include "ec_key_cert_marshal.h"
+
+/*****************************************************************************
+ * marshal_ec_pkey_to_der()
+ ****************************************************************************/
+int marshal_ec_pkey_to_der(EVP_PKEY ** ec_pkey_in,
+                           unsigned char **ec_der_bytes_out,
+                           int *ec_der_bytes_out_len)
+{
+  // validate that key to be marshalled is elliptic curve (EC) type
+  EVP_PKEY *pkey_ptr = *ec_pkey_in;
+
+  if (EVP_PKEY_base_id(pkey_ptr) != EVP_PKEY_EC)
+  {
+    const char *src_file_name = __FILE__;
+    const char *func_name = __func__;
+    int severity = 3;
+    const char *message = "PKEY to be marshalled is not of EC type";
+    const int line = __LINE__ + 1;
+    log_event_ocall(&src_file_name, &func_name, &line, &severity, &message);
+    return EXIT_FAILURE;
+  }
+
+  // Validate that a pointer to a NULL buffer pointer was passed in for
+  // the binary output byte array. Put it in that state if necessary.
+  // This will support memory allocation by the call to i2d_PrivateKey()
+  // for PKEY to DER format conversion rather than having to do the
+  // required size calculation and memory allocation beforehand.
+  if (*ec_der_bytes_out != NULL)
+  {
+    free(*ec_der_bytes_out);
+    *ec_der_bytes_out = NULL;
+  }
+
+  // A copy of the pointer is needed because i2d_PrivateKey() will
+  // modify the one it is passed as a parameter.
```

Comment:
Just for my clarity, it modifies the pointer but not the pointed-to-data?

Reply:
For the byte array, I believe that it increments the pointer you pass in and returns the incremented pointer (e.g., a value pointing to the end of the array). I was getting a lot of seg-faults when using these pointers later and I think that this was the reason. I do not know why I also made a copy of the pointers to the EVP_PKEY and X509 structs, so I just changed the code in these two marshal functions to use the passed in pointers to those structs directly.

Below are some excerpts from OpenSSL documentation related to this:

The ways that *ppin and *ppout are incremented after the operation can trap the unwary. See the WARNINGS section for some common errors. The reason for this-auto increment behaviour is to reflect a typical usage of ASN1 functions: after one structure is encoded or decoded another will be processed after it.

...

WARNINGS

Using a temporary variable is mandatory. A common mistake is to attempt to use a buffer directly as follows:

 int len;
 unsigned char *buf;

 len = i2d_X509(x, NULL);
 buf = OPENSSL_malloc(len);
 ...
 i2d_X509(x, &buf);
 ...
 OPENSSL_free(buf);
This code will result in buf apparently containing garbage because it was incremented after the call to point after the data just written. Also buf will no longer contain the pointer allocated by OPENSSL_malloc() and the subsequent call to OPENSSL_free() is likely to crash.

...